### PR TITLE
[DELETE] /api/groups/{groupId} // [GET] /api/groups/{groupId} // [GET] /api/groups

### PR DESCRIPTION
**Pull request by Wally5077**

### CODE REVIEW

File: Spring-Boot/Spring-Boot-Student/src/test/java/tw/waterball/judgegirl/springboot/student/controllers/it/GroupControllerIT.java [hunk 1]
```diff
@@ -47,16 +51,65 @@ public void WhenCreateGroupWithUniqueName_ShouldCreateSuccessfully() throws Exce
...
+    private ResultActions createGroup(String groupName) throws Exception {
+        CreateGroupUseCase.Request request = new CreateGroupUseCase.Request(groupName);
+        return mockMvc.perform(post(BASE_PATH)
                 .contentType(MediaType.APPLICATION_JSON)
                 .content(toJson(request)));
     }
 
+    @Test
+    public void WhenGetGroupByNonExistingGroupId_ShouldRespondNotFound() throws Exception {
+        int nonExistingStudentId = 123123;
```
Comment (Chaoyuuu) [inline Spring-Boot/Spring-Boot-Student/src/test/java/tw/waterball/judgegirl/springboot/student/controllers/it/GroupControllerIT.java:88]: This is `nonExistingGroupId`.

File: Spring-Boot/Spring-Boot-Student/src/test/java/tw/waterball/judgegirl/springboot/student/controllers/it/GroupControllerIT.java [hunk 2]
```diff
@@ -47,16 +51,65 @@ public void WhenCreateGroupWithUniqueName_ShouldCreateSuccessfully() throws Exce
...
+        getGroupById(groupId).andExpect(status().isNotFound());
+    }
+
+    private ResultActions getGroupById(Integer id) throws Exception {
+        return mockMvc.perform(get(BASE_PATH + "/{groupId}", id));
+    }
+
+    @Test
+    public void WhenDeleteGroupByNonExistingGroupId_ShouldRespondNotFound() throws Exception {
+        int nonExistingStudentId = 123123;
```
Comment (Chaoyuuu) [inline Spring-Boot/Spring-Boot-Student/src/test/java/tw/waterball/judgegirl/springboot/student/controllers/it/GroupControllerIT.java:106]: This is `nonExistingGroupId`.

File: Student-Service/src/main/java/tw/waterball/judgegirl/studentservice/domain/usecases/DeleteGroupByIdUseCase.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    public void execute(Integer groupId) throws NotFoundException {
+        validateGroupIdExist(groupId);
+        groupRepository.deleteGroupById(groupId);
+    }
+
+    private void validateGroupIdExist(Integer id) throws NotFoundException {
+        if (!groupRepository.existsById(id)) {
+            throw new NotFoundException();
+        }
+    }
```
Comment (Chaoyuuu) [inline Student-Service/src/main/java/tw/waterball/judgegirl/studentservice/domain/usecases/DeleteGroupByIdUseCase.java:27]: Why don't you throw `EmptyResultDataAccessException` ?
Then the code will look as following:
```java
public void execute(Integer groupId) throws NotFoundException {
        try {
              groupRepository.deleteGroupById(groupId);
        } catch (EmptyResultDataAccessException e) {
              throw new NotFoundException();
        }
    }
```

File: Student-Service/src/main/java/tw/waterball/judgegirl/studentservice/domain/usecases/GetGroupByIdUseCase.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    public void execute(Integer groupId, Presenter presenter) throws NotFoundException {
+        validateGroupIdExist(groupId);
+        presenter.setGroup(groupRepository.findGroupById(groupId));
+    }
+
+    private void validateGroupIdExist(Integer id) throws NotFoundException {
+        if (!groupRepository.existsById(id)) {
+            throw new NotFoundException();
+        }
+    }
```
Comment (Chaoyuuu) [inline Student-Service/src/main/java/tw/waterball/judgegirl/studentservice/domain/usecases/GetGroupByIdUseCase.java:28]: You can use `Optional` here !
```java
public void execute(Integer groupId, Presenter presenter) throws NotFoundException {
        presenter.setGroup(groupRepository.findGroupById(groupId)
                     .orElseThrow(NotFoundException::New));
    }
```
But the prerequisite is the `findGroupById` should return `Optional<Group>`.

**Review by Chaoyuuu — CHANGES_REQUESTED** [review]
Review Done

**Review by Chaoyuuu — APPROVED** [review]
LGTM. 😄

File: Spring-Boot/Spring-Boot-Student/src/main/java/tw/waterball/judgegirl/springboot/student/repositories/JpaGroupRepository.java [hunk 5]
```diff
@@ -34,4 +40,31 @@ public Group save(Group group) {
     public void deleteAll() {
         jpaGroupDataPort.deleteAll();
     }
+
+    @Override
+    public Optional<Group> findGroupById(Integer groupId) {
+        return jpaGroupDataPort.findById(groupId)
+                .map(groupData -> new Group(groupData.getId(), groupData.getName()));
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Student/src/main/java/tw/waterball/judgegirl/springboot/student/repositories/JpaGroupRepository.java:47]: `.map(groupData -> groupData.toEntity()));`

File: Spring-Boot/Spring-Boot-Student/src/main/java/tw/waterball/judgegirl/springboot/student/repositories/JpaGroupRepository.java [hunk 6]
```diff
@@ -34,4 +40,31 @@ public Group save(Group group) {
...
+
+    @Override
+    public boolean existsById(Integer id) {
+        return jpaGroupDataPort.existsById(id);
+    }
+
+    @Override
+    public List<Group> findAllGroup() {
+        return jpaGroupDataPort.findAll().stream()
+                .map(groupData -> new Group(groupData.getId(), groupData.getName()))
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Student/src/main/java/tw/waterball/judgegirl/springboot/student/repositories/JpaGroupRepository.java:58]: `.map(groupData -> groupData.toEntity()));`

File: Student-Service/src/main/java/tw/waterball/judgegirl/studentservice/domain/usecases/DeleteGroupByIdUseCase.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+/**
+ * @author - wally55077@gmail.com
+ */
+@Named
+@AllArgsConstructor
+public class DeleteGroupByIdUseCase {
+
+    private final GroupRepository groupRepository;
+
+    public void execute(Integer groupId) throws NotFoundException {
```
Comment (Johnny850807) [inline Student-Service/src/main/java/tw/waterball/judgegirl/studentservice/domain/usecases/DeleteGroupByIdUseCase.java:18]: `int groupId`

File: Student-Service/src/main/java/tw/waterball/judgegirl/studentservice/domain/usecases/GetGroupByIdUseCase.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+/**
+ * @author - wally55077@gmail.com
+ */
+@Named
+@AllArgsConstructor
+public class GetGroupByIdUseCase {
+
+    private final GroupRepository groupRepository;
+
+    public void execute(Integer groupId, Presenter presenter) throws NotFoundException {
```
Comment (Johnny850807) [inline Student-Service/src/main/java/tw/waterball/judgegirl/studentservice/domain/usecases/GetGroupByIdUseCase.java:19]: `int groupId`

File: Student-Service/src/main/java/tw/waterball/judgegirl/studentservice/domain/repositories/GroupRepository.java [hunk 9]
```diff
@@ -2,11 +2,22 @@
...
+
 public interface GroupRepository {
 
     boolean existsByName(String name);
 
     Group save(Group group);
 
     void deleteAll();
+
+    Optional<Group> findGroupById(Integer groupId);
```
Comment (Johnny850807) [inline Student-Service/src/main/java/tw/waterball/judgegirl/studentservice/domain/repositories/GroupRepository.java:16]: `int groupId`

File: Student-Service/src/main/java/tw/waterball/judgegirl/studentservice/domain/repositories/GroupRepository.java [hunk 10]
```diff
@@ -2,11 +2,22 @@
...
+
+    boolean existsById(Integer id);
```
Comment (Johnny850807) [inline Student-Service/src/main/java/tw/waterball/judgegirl/studentservice/domain/repositories/GroupRepository.java:18]: `int groupId`

File: Student-Service/src/main/java/tw/waterball/judgegirl/studentservice/domain/repositories/GroupRepository.java [hunk 11]
```diff
@@ -2,11 +2,22 @@
...
+
+    List<Group> findAllGroup();
+
+    void deleteGroupById(Integer groupId);
```
Comment (Johnny850807) [inline Student-Service/src/main/java/tw/waterball/judgegirl/studentservice/domain/repositories/GroupRepository.java:22]: `groupId`

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
Great code, just some minor issues!

**Review by Johnny850807 — APPROVED** [review]
LGTM

**Pull request merged**